include/deal.II/matrix_free/cuda_matrix_free.templates.h

```diff
@@ -0,0 +1,709 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+#ifndef dealii__cuda_matrix_free_templates_h
+#define dealii__cuda_matrix_free_templates_h
+
+#include <deal.II/matrix_free/cuda_matrix_free.h>
+
+#ifdef DEAL_II_WITH_CUDA
+
+#include <deal.II/base/graph_coloring.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/grid/filtered_iterator.h>
+#include <deal.II/matrix_free/shape_info.h>
+#include <cuda_runtime_api.h>
+#include <functional>
+
+#define BLOCK_SIZE 128
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace CUDAWrappers
+{
+  namespace internal
+  {
+    // These variables are stored in the device constant memory.
+    // TODO: use a template parameter instead of a macro
+#define MAX_ELEM_DEGREE 10
+    __constant__ double global_shape_values[(MAX_ELEM_DEGREE+1) * (MAX_ELEM_DEGREE+1)];
+    __constant__ double global_shape_gradients[(MAX_ELEM_DEGREE+1) * (MAX_ELEM_DEGREE+1)];
+
+    template <typename Number>
+    using CUDAVector = ::dealii::LinearAlgebra::CUDAWrappers::Vector<Number>;
+
+    // These functions are not part of any class.
+    // TODO find a better name than T
+    template <typename T>
+    void transpose(const unsigned int N, const unsigned M, const T *src, T *dst)
+    {
+      // src is N X M
+      // dst is M X N
+      for (unsigned int i=0; i<N; ++i)
+        for (unsigned int j=0; j<M; ++j)
+          dst[j*N+i] = src[i*M+j];
+    }
+
+
+
+    template <typename T>
+    void transpose_in_place(std::vector<T> &array_host,
```

Comment:
i know it's WIP, but here also some internal functions/classes are not documented.